egui/src/widgets/text_edit.rs

```diff
@@ -656,7 +660,18 @@ impl<'t> TextEdit<'t> {
 
         ui.memory().id_data.insert(id, state);
 
-        response.widget_info(|| WidgetInfo::text_edit(&*text));
+        if response.changed {
+            response.widget_info(|| WidgetInfo::text_edit(&*prev_text, &*text));
+        } else if let Some(text_cursor) = text_cursor {
+            let char_range =
+                text_cursor.primary.ccursor.index..=text_cursor.secondary.ccursor.index;
+            let info = WidgetInfo::text_selection_changed(char_range, &*text);
+            response
+                .ctx
+                .output()
+                .events
+                .push(OutputEvent::TextSelectionChanged(info));
+        }
```

Comment:
Why the focus event aren't being sent anymore:

It is the call to `response.widget_info(…)` that sends those events (if applicable) and you now only call it when the text has changed.

So it will work again with something like:

``` rust
let text_cursor_changed = old_text_cursor != text_cursor;

if text_was_edited {
    let info = WidgetInfo::text_changed(prev_text, text);
    self.ctx.output().events.push(OutputEvent::TextChanged(info));
} else if text_cursor_changed && text_cursor.is_some() {
    let text_cursor = text_cursor.unwrap();
    let char_range = text_cursor.primary.ccursor.index..text_cursor.secondary.ccursor.index;
    let info = WidgetInfo::text_cursor_changed(char_range, text);
    self.ctx.output().events.push(OutputEvent::TextCursorChanged(info));
}  else {
    // This will emit events on focus, click, etc:
    response.widget_info(|| WidgetInfo::text_edit(&*text));
}
```

the reasoning for putting it in the else is that we don't want to send two events in the same frame for the same widget.